File: frontend/src/entities/template/lib/grammar.test.ts
/// <reference types="node" />
import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { describe, expect, it } from 'vitest'
import {
  askFields,
  decode,
  parse,
  parseTemplate,
  serialize,
  templateAskFields,
  type TemplateNode,
} from './grammar'

// The node reference is local to this file on purpose: the app tsconfig deliberately does not
// expose node types, so app code cannot reach the filesystem. Reading the shared fixture is a
// test-only need, and copying the fixture into the frontend would defeat the point of it.

/** The SAME fixture file the Go parser's suite reads. Two implementations of one grammar stay
 *  honest only if a new rule has one place to land (spec/legacy/tech/post-template-grammar.md §4). */
interface FixtureNode {
  t: string
  raw?: string
  text?: string
  kind?: string
  label?: string
  count?: number
  each?: string
  children?: FixtureNode[]
}
interface FixtureCase {
  name: string
  /** "" when the case has no title area (TMPL-50). */
  titleArea?: string
  body: string
  titleNodes?: FixtureNode[]
  nodes?: FixtureNode[]
  /** `area` is absent for a refusal in the body. */
  error?: { line: number; reason: string; area?: string }
}

const fixturePath = resolve(
  import.meta.dirname,
  '../../../../../backend/internal/template/testdata/grammar/cases.json',
)
const fixture = JSON.parse(readFileSync(fixturePath, 'utf8'))
const cases: FixtureCase[] = fixture.cases
/** The ceilings the FIXTURE declares, never this deployment's own: a `count` or a data-field
 *  case has to mean the same thing on both sides, which is exactly what this file guarantees. */
const options = {
  photoRowMax: fixture.photoRowMax as number,
  askMaxPerBody: fixture.askMaxPerBody as number,
}

function expectNodes(got: readonly TemplateNode[], want: readonly FixtureNode[], path: string) {
  expect(
    got.map((n) => n.kind),
    `${path}: kinds`,
  ).toEqual(want.map((n) => n.t))
  want.forEach((expected, i) => {
    const actual = got[i]
    const at = `${path}[${i}]`
    if (expected.t === 'literal') expect(actual.text, at).toBe(expected.raw)
    if (expected.t === 'write' || expected.t === 'note') {
      expect(decode(actual.text ?? ''), at).toBe(expected.text)
    }
    if (expected.t === 'slot') {
      expect(actual.slotKind, at).toBe(expected.kind)
      expect(decode(actual.label ?? ''), at).toBe(expected.label)
      expect(actual.count, at).toBe(expected.count)
    }
    if (expected.t === 'ask') {
      expect(decode(actual.label ?? ''), at).toBe(expected.label)
      expect(decode(actual.text ?? ''), at).toBe(expected.text)
    }
    if (expected.t === 'repeat') {
      expect(actual.each, at).toBe(expected.each)
      expectNodes(actual.children ?? [], expected.children ?? [], at)
    }
  })
}

describe('template grammar against the shared fixtures', () => {
  it('runs the ceilings the fixture declares', () => {
    expect(options.photoRowMax).toBeGreaterThan(0)
    expect(options.askMaxPerBody).toBeGreaterThan(0)
  })

  it('reads at least one accepted and one refused case', () => {
    expect(cases.filter((c) => c.nodes).length).toBeGreaterThan(0)
    expect(cases.filter((c) => c.error).length).toBeGreaterThan(0)
  })

  it('reads at least one title-area case', () => {
    expect(cases.filter((c) => (c.titleArea ?? '') !== '').length).toBeGreaterThan(0)
  })

  // Every case is a (title area, body) pair parsed as one template. One with no title area must
  // also give the body-only parse's identical verdict, which keeps every template saved before
  // title areas existed exactly where it was.
  cases.forEach((testCase) => {
    it(testCase.name, () => {
      const titleArea = testCase.titleArea ?? ''
      const result = parseTemplate(titleArea, testCase.body, options)
      if (titleArea === '') {
        const bodyOnly = parse(testCase.body, options)
        expect(bodyOnly.ok).toBe(result.ok)
        if (!bodyOnly.ok && !result.ok) expect(bodyOnly.failure).toEqual(result.failure)
        if (bodyOnly.ok && result.ok)
          expect(serialize(bodyOnly.nodes)).toBe(serialize(result.nodes))
      }
      if (testCase.error) {
        expect(result.ok, 'expected a parse failure').toBe(false)
        if (result.ok) return
        expect(result.failure).toEqual({ area: 'body', ...testCase.error })
        return
      }
      expect(result.ok, `unexpected failure: ${JSON.stringify(result)}`).toBe(true)
      if (!result.ok) return
      expectNodes(result.titleNodes, testCase.titleNodes ?? [], 'titleNodes')
      expectNodes(result.nodes, testCase.nodes ?? [], 'nodes')
      // Every accepted area must serialize back byte for byte: this is the round-trip
      // guarantee the builder's source toggle rests on (change 25 AC8).
      expect(serialize(result.titleNodes)).toBe(titleArea)
      expect(serialize(result.nodes)).toBe(testCase.body)
    })
  })

  it('names the area a failure sits in, a body on its own included', () => {
    const body = parse('<writer>', options)
    expect(body.ok ? null : body.failure.area).toBe('body')
    const title = parse('<note>톤</note>', { ...options, titleArea: true })
    expect(title.ok ? null : title.failure).toEqual({
      line: 1,
      reason: 'not_in_title',
      area: 'title_area',
    })
    expect(parse('<slot kind="photo"/>', options).ok).toBe(true)
  })
})

/** The differential corpus: 600 machine-generated bodies plus the GO parser's verdict on each.
 *
 *  The hand-written fixtures above pin the rules anyone thought to write down. This pins the
 *  shapes nobody did — the combinations a mistake in either implementation actually falls into.
 *  Regenerate with `TEMPLATE_CORPUS_REGEN=1 go test ./internal/template/ -run TestCorpus`. */
interface CorpusCase {
  body: string
  ok: boolean
  line?: number
  reason?: string
}

interface TitleCorpusCase {
  titleArea: string
  body: string
  ok: boolean
  line?: number
  reason?: string
  area?: string
}

const corpusFile = JSON.parse(
  readFileSync(
    resolve(
      import.meta.dirname,
      '../../../../../backend/internal/template/testdata/grammar/corpus.json',
    ),
    'utf8',
  ),
)
const corpus: CorpusCase[] = corpusFile.cases
/** 200 generated (title area, body) pairs with the Go parser's verdict (TMPL-50). */
const titleCorpus: TitleCorpusCase[] = corpusFile.titleCases

describe('the TypeScript parser agrees with the Go parser', () => {
  it('reaches the same verdict, line and reason on every corpus body', () => {
    const disagreements: string[] = []
    for (const testCase of corpus) {
      const result = parse(testCase.body, options)
      if (result.ok !== testCase.ok) {
        disagreements.push(
          `${JSON.stringify(testCase.body)}: go ${testCase.ok ? 'accepted' : 'refused'}, ts ${
            result.ok ? 'accepted' : 'refused'
          }`,
        )
        continue
      }
      if (!result.ok) {
        if (result.failure.line !== testCase.line || result.failure.reason !== testCase.reason) {
          disagreements.push(
            `${JSON.stringify(testCase.body)}: go ${testCase.reason}@${testCase.line}, ts ${
              result.failure.reason
            }@${result.failure.line}`,
          )
        }
        continue
      }
      if (serialize(result.nodes) !== testCase.body) {
        disagreements.push(`${JSON.stringify(testCase.body)}: ts round trip changed the body`)
      }
    }
    expect(disagreements.slice(0, 10)).toEqual([])
  })

  it('reaches the same verdict, line, reason and area on every title-area pair', () => {
    expect(titleCorpus.length).toBe(200)
    const disagreements: string[] = []
    for (const testCase of titleCorpus) {
      const pair = `${JSON.stringify(testCase.titleArea)} | ${JSON.stringify(testCase.body)}`
      const result = parseTemplate(testCase.titleArea, testCase.body, options)
      if (result.ok !== testCase.ok) {
        disagreements.push(
          `${pair}: go ${testCase.ok ? 'accepted' : 'refused'}, ts ${result.ok ? 'accepted' : 'refused'}`,
        )
        continue
      }
      if (!result.ok) {
        const { line, reason, area } = result.failure
        if (line !== testCase.line || reason !== testCase.reason || area !== testCase.area) {
          disagreements.push(
            `${pair}: go ${testCase.reason}@${testCase.area}:${testCase.line}, ts ${reason}@${area}:${line}`,
          )
        }
        continue
      }
      if (
        serialize(result.titleNodes) !== testCase.titleArea ||
        serialize(result.nodes) !== testCase.body
      ) {
        disagreements.push(`${pair}: ts round trip changed the pair`)
      }
    }
    expect(disagreements.slice(0, 10)).toEqual([])
  })
})

/** What the write screen reads off a body (TEMPLATE-43). It is the one place ① learns which
 *  fields exist, so it has to answer for a body nobody can fix from there. */
describe('the data fields a body asks for', () => {
  it('lists them in body order with the flavor each one feeds', () => {
    expect(
      askFields(
        '오늘의 기록\n<ask label="방문일"/>\n<slot kind="photo"/>\n<ask label="총평">총평을 쓰세요</ask>',
        options,
      ),
    ).toEqual([
      { label: '방문일', flavor: 'verbatim' },
      { label: '총평', flavor: 'write' },
    ])
  })

  it('decodes a title the way the builder shows it', () => {
    expect(askFields('<ask label="네이버 &quot;별점&quot;"/>', options)).toEqual([
      { label: '네이버 "별점"', flavor: 'verbatim' },
    ])
  })

  it('asks for nothing when the body has no field', () => {
    expect(askFields('<write>인트로</write>', options)).toEqual([])
  })

  // ① is not where a broken template is fixed: it renders no fields rather than an error.
  it('asks for nothing when the body does not parse', () => {
    expect(askFields('<ask label="총평"/>\n<ask label="총평"/>', options)).toEqual([])
    expect(askFields('<writer>', options)).toEqual([])
  })
})

/** What ① will read off a template once it has a title area: one namespace, title first. */
describe('the data fields a template asks for', () => {
  it('lists the fields of the title area first, then those of the body', () => {
    expect(
      templateAskFields(
        '맛집 | <ask label="가게 이름"/>',
        '<write>인트로</write>\n<ask label="총평">총평을 쓰세요</ask>',
        options,
      ),
    ).toEqual([
      { label: '가게 이름', flavor: 'verbatim' },
      { label: '총평', flavor: 'write' },
    ])
  })

  it('asks for nothing when the title area does not parse', () => {
    expect(templateAskFields('<note>톤</note>', '<ask label="총평"/>', options)).toEqual([])
  })

  it('asks for nothing when the body does not parse', () => {
    expect(templateAskFields('<ask label="가게"/>', '<writer>', options)).toEqual([])
  })

  it('asks for nothing when the two areas reuse a title', () => {
    expect(templateAskFields('<ask label="가게"/>', '<ask label="가게"/>', options)).toEqual([])
  })

  it('reads an empty title area as the body alone', () => {
    const body = '오늘의 기록\n<ask label="방문일"/>\n<ask label="총평">총평을 쓰세요</ask>'
    expect(templateAskFields('', body, options)).toEqual(askFields(body, options))
  })
})

File: backend/internal/template/grammar_test.go
package template

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// The fixture shape is shared with the frontend parser. Both suites read the SAME file, so a
// grammar rule cannot land on one side only (spec/legacy/tech/post-template-grammar.md §4).
type fixtureNode struct {
	T        string        `json:"t"`
	Raw      string        `json:"raw"`
	Text     string        `json:"text"`
	Kind     string        `json:"kind"`
	Label    string        `json:"label"`
	Count    int           `json:"count"`
	Each     string        `json:"each"`
	Children []fixtureNode `json:"children"`
}

// fixtureParseOptions is the ceilings the SHARED fixture declares (`photoRowMax`,
// `askMaxPerBody`), not the ones this deployment configured. A fixture that read the
// environment would pass or fail depending on where it ran, and the TypeScript harness could
// not reproduce it at all.
var fixtureParseOptions = ParseOptions{PhotoRowMax: 4, AskMaxPerBody: 3}

type fixtureCase struct {
	Name string `json:"name"`
	// TitleArea is "" when the case has none (TMPL-50).
	TitleArea  string        `json:"titleArea"`
	Body       string        `json:"body"`
	TitleNodes []fixtureNode `json:"titleNodes"`
	Nodes      []fixtureNode `json:"nodes"`
	Error      *struct {
		Line   int    `json:"line"`
		Reason string `json:"reason"`
		// Area is where the refusal sits; absent means the body.
		Area string `json:"area"`
	} `json:"error"`
}

func loadFixtures(t *testing.T) []fixtureCase {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("testdata", "grammar", "cases.json"))
	if err != nil {
		t.Fatal(err)
	}
	var file struct {
		PhotoRowMax   int           `json:"photoRowMax"`
		AskMaxPerBody int           `json:"askMaxPerBody"`
		Cases         []fixtureCase `json:"cases"`
	}
	if err := json.Unmarshal(raw, &file); err != nil {
		t.Fatal(err)
	}
	if len(file.Cases) == 0 {
		t.Fatal("grammar fixtures are empty")
	}
	// The two harnesses must run the same ceiling or a `count` case means different things
	// on the two sides, which is the exact drift this file exists to prevent.
	if file.PhotoRowMax != fixtureParseOptions.PhotoRowMax {
		t.Fatalf("fixture photoRowMax = %d, want %d", file.PhotoRowMax, fixtureParseOptions.PhotoRowMax)
	}
	if file.AskMaxPerBody != fixtureParseOptions.AskMaxPerBody {
		t.Fatalf("fixture askMaxPerBody = %d, want %d", file.AskMaxPerBody, fixtureParseOptions.AskMaxPerBody)
	}
	return file.Cases
}

// Every case is a (title area, body) pair parsed as one template. A case with no title area
// must also give the body-only Parse's identical verdict, which is what keeps every template
// saved before title areas existed exactly where it was.
func TestParseAgainstSharedFixtures(t *testing.T) {
	for _, tc := range loadFixtures(t) {
		t.Run(tc.Name, func(t *testing.T) {
			title, nodes, err := ParseTemplate(tc.TitleArea, tc.Body, fixtureParseOptions)
			if tc.TitleArea == "" {
				bodyOnly, bodyErr := Parse(tc.Body, fixtureParseOptions)
				if !sameVerdict(nodes, err, bodyOnly, bodyErr) {
					t.Fatalf("ParseTemplate with no title area = (%v, %v), Parse = (%v, %v)", kinds(nodes), err, kinds(bodyOnly), bodyErr)
				}
			}
			if tc.Error != nil {
				if err == nil {
					t.Fatalf("expected %s on line %d, parsed %d title and %d body nodes", tc.Error.Reason, tc.Error.Line, len(title), len(nodes))
				}
				parseErr, ok := err.(*ParseError)
				if !ok {
					t.Fatalf("error %v is not a ParseError", err)
				}
				wantArea := tc.Error.Area
				if wantArea == "" {
					wantArea = AreaBody
				}
				if parseErr.Reason != tc.Error.Reason || parseErr.Line != tc.Error.Line || parseErr.Area != wantArea {
					t.Fatalf("got %s on %s line %d, want %s on %s line %d",
						parseErr.Reason, parseErr.Area, parseErr.Line, tc.Error.Reason, wantArea, tc.Error.Line)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected parse error: %v", err)
			}
			assertNodes(t, title, tc.TitleNodes, "titleNodes")
			assertNodes(t, nodes, tc.Nodes, "")
			// Every accepted area must serialize back byte-for-byte: this is the round-trip
			// guarantee the builder's 원문 toggle rests on (change 25 AC8).
			if round := Serialize(title); round != tc.TitleArea {
				t.Fatalf("round trip changed the title area:\n got %q\nwant %q", round, tc.TitleArea)
			}
			if round := Serialize(nodes); round != tc.Body {
				t.Fatalf("round trip changed the body:\n got %q\nwant %q", round, tc.Body)
			}
		})
	}
}

// sameVerdict compares two parses by acceptance, node kinds and the error each refused with.
func sameVerdict(nodes []Node, err error, otherNodes []Node, otherErr error) bool {
	if (err == nil) != (otherErr == nil) {
		return false
	}
	if err != nil {
		first, firstOK := err.(*ParseError)
		second, secondOK := otherErr.(*ParseError)
		return firstOK && secondOK && *first == *second
	}
	return Serialize(nodes) == Serialize(otherNodes) && len(kinds(nodes)) == len(kinds(otherNodes))
}

// A body parsed on its own names the body, and one parsed as a title area names the title.
func TestParseNamesTheAreaItParsed(t *testing.T) {
	for _, body := range []string{"<writer>", "<write>메뉴", "<ask label=\"a\"/><ask label=\"a\"/>", "</write>"} {
		_, err := Parse(body, fixtureParseOptions)
		parseErr, ok := err.(*ParseError)
		if !ok || parseErr.Area != AreaBody {
			t.Errorf("Parse(%q) error = %#v, want the body", body, err)
		}
		titleOpts := fixtureParseOptions
		titleOpts.TitleArea = true
		_, err = Parse(body, titleOpts)
		if parseErr, ok := err.(*ParseError); !ok || parseErr.Area != AreaTitle {
			t.Errorf("Parse(%q) as a title area error = %#v, want the title area", body, err)
		}
	}
	if _, err := Parse(`<slot kind="photo"/>`, fixtureParseOptions); err != nil {
		t.Fatalf("a body still admits a photo position: %v", err)
	}
}

func assertNodes(t *testing.T, got []Node, want []fixtureNode, path string) {
	t.Helper()
	if len(got) != len(want) {
		t.Fatalf("%s: got %d nodes, want %d (%v)", path, len(got), len(want), kinds(got))
	}
	for i := range want {
		at := path + "[" + itoa(i) + "]"
		if string(got[i].Kind) != want[i].T {
			t.Fatalf("%s: kind %q, want %q", at, got[i].Kind, want[i].T)
		}
		switch want[i].T {
		case "literal":
			if got[i].Text != want[i].Raw {
				t.Fatalf("%s: literal %q, want %q", at, got[i].Text, want[i].Raw)
			}
		case "write", "note":
			if Decode(got[i].Text) != want[i].Text {
				t.Fatalf("%s: text %q, want %q", at, Decode(got[i].Text), want[i].Text)
			}
		case "slot":
			if string(got[i].SlotKind) != want[i].Kind {
				t.Fatalf("%s: slot kind %q, want %q", at, got[i].SlotKind, want[i].Kind)
			}
			if Decode(got[i].Label) != want[i].Label {
				t.Fatalf("%s: slot label %q, want %q", at, Decode(got[i].Label), want[i].Label)
			}
			if got[i].Count != want[i].Count {
				t.Fatalf("%s: slot count %d, want %d", at, got[i].Count, want[i].Count)
			}
		case "ask":
			if Decode(got[i].Label) != want[i].Label {
				t.Fatalf("%s: ask label %q, want %q", at, Decode(got[i].Label), want[i].Label)
			}
			if Decode(got[i].Text) != want[i].Text {
				t.Fatalf("%s: ask text %q, want %q", at, Decode(got[i].Text), want[i].Text)
			}
		case "repeat":
			if got[i].Each != want[i].Each {
				t.Fatalf("%s: each %q, want %q", at, got[i].Each, want[i].Each)
			}
			assertNodes(t, got[i].Children, want[i].Children, at)
		}
	}
}

func kinds(nodes []Node) []NodeKind {
	out := make([]NodeKind, 0, len(nodes))
	for _, node := range nodes {
		out = append(out, node.Kind)
	}
	return out
}

func itoa(value int) string {
	if value == 0 {
		return "0"
	}
	digits := ""
	for value > 0 {
		digits = string(rune('0'+value%10)) + digits
		value /= 10
	}
	return digits
}
